src/parser/mod.rs

```diff
@@ -475,6 +475,12 @@ impl<'a> Parser<'a> {
                     if expecting_statement_delimiter && word.keyword == Keyword::END {
                         break;
                     }
+                    // Treat batch delimiter as an end of statement
+                    if expecting_statement_delimiter && dialect_of!(self is MsSqlDialect) {
+                        if let Some(Statement::Go { count: _ }) = stmts.last() {
+                            expecting_statement_delimiter = false;
+                        }
+                    }
```

Comment:
not sure I followed this part, what was the intention?

Reply:
I tried to explain this in the commit message & the "multiple_gos" example in the test. Basically, we want this to parse as 4 statements:

```mssql
SELECT 1;
GO
SELECT 2;
GO
```

The second `GO` is terminated by EOF, so no problem. But the first `GO` _cannot_ end with a semi colon (This overlaps with the ongoing "no semi-colons for statements" discussion but from a different angle). So we need to figure out how to parse it all out without the existing conventional semi-colon logic.

What I came up with is to basically just look for a GO and have that signal that we _aren't_ then also looking for statement delimiter (eg, semi-colon). This also makes sense semantically because statements can't extend between batches, so any batch delimiter also functions as a statement delimiter.

Reply:
In terms of the code, I worked out a perhaps slightly less confusing way to set whether or not a semi-colon is expected, [here](https://github.com/apache/datafusion-sqlparser-rs/pull/1808/files#diff-fc6a8d66b8cb6bd48a119a70e489cbcef82e7f551e7cf08e8e972ef4e774ef49R487-R493).

This branch can be updated accordingly if we agree over there

Reply:
I've now done that due to another problem I noticed, for IF statements (new test case also added...)